### PR TITLE
Adding TSLint rules

### PR DESCRIPTION
**Pull request by KamranAsif**

### CODE REVIEW

File: src/shallowClone.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export const shallowClone = <T extends object>(source: T): T => {
+  const base = Array.isArray(source) ? [] : {};
+  // TODO(asif): Remove this hack when TS/pull/13288 is merged in
+  const sourceObj = source as {} as object;
```
Comment (wcjordan) [inline src/shallowClone.ts]: what's the syntax equivalent to in JS?  I'm not familiar w/ it

File: src/walkTree.ts [hunk 2]
```diff
@@ -16,15 +19,15 @@ const walkTree = <T>(source: T, revision: T): T => {
...
+  if (!isObject(source) || !isObject(revision)) {
     return revision;
   }
 
   // Gather keys. For arrays, its just a list of indexes
   // TODO (asif) See if it'll be quicker walk arrays instead of getting its keys
   type keyType = keyof T;
   const sourceKeys = Object.keys(source) as keyType[];
-  const revisionKeys = Object.keys(revision) as keyType[];
+  // const revisionKeys = Object.keys(revision) as keyType[];
```
Comment (wcjordan) [inline src/walkTree.ts]: unneeded?

File: test/retainFn-spec.ts [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import retainFn from '../src/retainFn';
+import {retainFn} from '../src/retainFn';
```
Comment (wcjordan) [inline test/retainFn-spec.ts]: nit spacing in braces for style

File: src/shallowClone.ts (hunk 1 above)
Comment (KamranAsif) [reply src/shallowClone.ts]: It's just const sourceObj = source;
I had to add the as {} as object to make it 'forget' its typing until that bug is resolved.

File: src/walkTree.ts (hunk 2 above)
Comment (KamranAsif) [reply src/walkTree.ts]: Yeah, but I used it in the next CL; so just ignore.

File: src/shallowClone.ts (hunk 1 above)
Comment (wcjordan) [reply src/shallowClone.ts]: ah cool thanks

File: test/retainFn-spec.ts (hunk 3 above)
Comment (KamranAsif) [reply test/retainFn-spec.ts]: Hmm this is actually the style I want, to make goog standards. 
I'll update the rule to disallow spacing in brackets.

**Pull request merged**